File: scripts/embed.py
"""Embedding generation for the ingestion pipeline.

Local, offline, deterministic embeddings via sentence-transformers using the
all-MiniLM-L6-v2 model (384-dim, mean pooling + L2 normalization). The PyTorch
checkpoint here (sentence-transformers/all-MiniLM-L6-v2) shares the SAME weights
as the ONNX checkpoint used at runtime by the browser Transformers.js
implementation (Xenova/all-MiniLM-L6-v2, lib/embed.ts), so claim vectors computed
here are comparable to query vectors computed in the browser.

Model weights are cached under scripts/.cache on first run (a one-time download).
"""
from __future__ import annotations

from pathlib import Path

EMBEDDING_MODEL = "sentence-transformers/all-MiniLM-L6-v2"
_CACHE_DIR = Path(__file__).resolve().parent / ".cache"

_model = None


def _get_model():
    """Lazily load the sentence-transformers model once per process."""
    global _model
    if _model is None:
        from sentence_transformers import SentenceTransformer

        # cache_folder keeps weights local/reproducible across runs
        _model = SentenceTransformer(
            EMBEDDING_MODEL,
            cache_folder=str(_CACHE_DIR),
        )
    return _model


def embed_texts(texts: list[str]) -> list[list[float]]:
    """Embed a batch of texts into normalized 384-dim vectors.

    SentenceTransformer defaults to mean pooling and L2-normalizes embeddings
    when normalize_embeddings=True, matching the browser implementation.
    """
    model = _get_model()
    if not texts:
        return []
    vectors = model.encode(texts, normalize_embeddings=True)
    return [list(map(float, v)) for v in vectors]

File: lib/embed.ts
// Browser embedding via Transformers.js using the SAME weights as the Python
// ingestion pipeline (scripts/embed.py): all-MiniLM-L6-v2, 384-dim, mean
// pooling + L2 normalization. Query vectors computed here are comparable to
// stored claim vectors at query time.
//
// The transformers module is imported dynamically so it is never evaluated
// during server-side prerendering (it constructs WASM URLs that fail in Node).
export const EMBEDDING_MODEL = 'Xenova/all-MiniLM-L6-v2';
export const EMBEDDING_DIM = 384;

type FeatureExtractor = (
  texts: string[],
  opts: { pooling: 'mean'; normalize: boolean },
) => Promise<{ tolist(): number[][] }>;

let extractorPromise: Promise<FeatureExtractor> | null = null;

async function getExtractor(): Promise<FeatureExtractor> {
  if (!extractorPromise) {
    const { pipeline } = await import('@huggingface/transformers');
    extractorPromise = pipeline('feature-extraction', EMBEDDING_MODEL) as Promise<
      FeatureExtractor
    >;
  }
  return extractorPromise;
}

export async function embedTexts(texts: string[]): Promise<number[][]> {
  if (texts.length === 0) return [];
  const extractor = await getExtractor();
  // pooling 'mean' + normalize true matches sentence-transformers defaults.
  const output = await extractor(texts, { pooling: 'mean', normalize: true });
  return output.tolist();
}

